Public/Src/Pips/Dll/Artifacts/PipArtifacts.cs

```diff
@@ -230,5 +235,31 @@ public static bool ForEachOutput(Pip pip, Func<FileOrDirectoryArtifact, bool> ou
 
             return result;
         }
+
+        /// <summary>
+        /// Gets the input files consumed by a pip, enumerate the input directories recuresively to get all files.
+        /// </summary>
+        public static IReadOnlyCollection<AbsolutePath> GetAllInputs(Process process, PathTable pathTable)
+        {
+            var inputs = new HashSet<AbsolutePath>();
+
+            inputs.AddRange(process.Dependencies.Select(d => d.Path));
+
+            foreach (var inDir in process.DirectoryDependencies)
+            {
+                FileUtilities.EnumerateDirectoryEntries(

```

Comment:
>EnumerateDirectoryEntries [](start = 30, length = 25)

I don't think this is correct. Pips don't consume every file in their directory dependencies.